### PR TITLE
Serialisation in the primary HDU of the Dataset name

### PR DESCRIPTION
**Pull request by bkhelifi**
**Description**
Based on the discussion described in the PR #4045, this pull request is one of the three aiming to deal correctly with the serialization of the `Dataset` names. This PR is then associated to the issue #3750.

This pull request is about the serialization of the dataset names with the FITS file in the header of the primary HDU

### CODE REVIEW

File: gammapy/datasets/map.py [hunk 1]
```diff
@@ -1230,7 +1234,12 @@ def write(self, filename, overwrite=False):
 
     @classmethod
     def _read_lazy(cls, name, filename, cache, format=format):
-        kwargs = {"name": name}
+        ds_name = make_name(name)
+        hdulist = fits.open(str(make_path(filename)))
```
Comment (adonath) [inline gammapy/datasets/map.py:1238]: As you open the hdulist here anyway there is not advantage in putting it in `_read_lazy`. So the extortion of the name could just always happen in `MapDataset.read()`

Comment (adonath) [reply gammapy/datasets/map.py:1238]: Just remove the modification here...

File: gammapy/datasets/map.py [hunk 2]
```diff
@@ -1289,15 +1298,17 @@ def read(cls, filename, name=None, lazy=False, cache=True, format="gadf"):
...
-        name = make_name(name)
 
         if lazy:
             return cls._read_lazy(
                 name=name, filename=filename, cache=cache, format=format
             )
         else:
             with fits.open(str(make_path(filename)), memmap=False) as hdulist:
-                return cls.from_hdulist(hdulist, name=name, format=format)
+                ds_name = make_name(name)
```
Comment (adonath) [inline gammapy/datasets/map.py:1308]: Again just remove the modification here...

File: gammapy/datasets/map.py [hunk 3]
```diff
@@ -1289,15 +1298,17 @@ def read(cls, filename, name=None, lazy=False, cache=True, format="gadf"):
         dataset : `MapDataset`
             Map dataset.
         """
-        name = make_name(name)
```
Comment (adonath) [inline gammapy/datasets/map.py:1301]:
```suggestion
if name is None:
    header = fits.getheader(filename)
    name = header.get("NAME", name)
name = make_name(name)
```

**Review by adonath — CHANGES_REQUESTED** [review]
Thanks @bkhelifi I have left some minor comments for simplification...

File: gammapy/datasets/map.py (hunk 1 above)
Comment (bkhelifi) [reply gammapy/datasets/map.py:1238]: In the read function, the name management is made for the non-lazy method. In the case of the lazy one, why we should not manage it?

File: gammapy/datasets/tests/test_map.py [hunk 4]
```diff
@@ -1147,8 +1150,10 @@ def test_map_dataset_on_off_fits_io(images, lazy, tmp_path):
     if lazy:
         with pytest.raises(NotImplementedError):
             dataset_new = MapDatasetOnOff.read(tmp_path / "test.fits", lazy=lazy)
+            assert dataset_new.name == 'MapDatasetOnOff-test'
```
Comment (registerrier) [inline gammapy/datasets/tests/test_map.py:1153]: Should we expect something after an error is raised?

**Review by registerrier — DISMISSED** [review]
Thanks @bkhelifi . This looks good. See minor inline comment.

One general question though. The ogip format for `SpectrumOnOffDataset.write` sets the  `dataset.name` to the `OBS_ID` keyword and `.read` takes the name from the `OBS_ID` not the `NAME` keyword. Since the two formats are independent that is probably not an issue, but this will require some documentation.

**Review by adonath — APPROVED** [review]
Thanks @bkhelifi, no further comments from my side!

**Review by registerrier — APPROVED** [review]
Thanks @bkhelifi 

No comments from my side.

**Pull request merged**